cruise-control/src/main/java/com/linkedin/kafka/cruisecontrol/detector/BrokerFailureDetector.java

```diff
@@ -88,16 +88,22 @@ void startDetection() {
     _zkClient.subscribeChildChanges(ZkUtils.BrokerIdsPath(), new BrokerFailureListener());
   }
 
-  synchronized void detectBrokerFailures() {
+  private synchronized void detectBrokerFailures(Set<Integer> aliveBrokers) {
     // update the failed broker information based on the current state.
-    updateFailedBrokerList(aliveBrokers());
-    // persist the updated failed broker list.
-    persistFailedBrokerList();
+    boolean updated = updateFailedBrokers(aliveBrokers);
+    if (updated) {
+      // persist the updated failed broker list.
+      persistFailedBrokerList();
+    }
     // Report the failures to anomaly detector to handle.
     reportBrokerFailures();
   }
 
-  Map<Integer, Long> failedBrokers() {
+  synchronized void detectBrokerFailures() {
+    detectBrokerFailures(aliveBrokers());
+  }
+
+  synchronized Map<Integer, Long> failedBrokers() {
```

Comment:
This method is called by `reportBrokerFailures()` , `reportBrokerFailures()` is called by `detectBrokerFailures()` which is already protected by `synchronized`, maybe this `synchronized` is not necessary?

Reply:
`_failedBrokers` is a shared resource. While today the `failedBrokers()` function is called from within the described synchronized code path, since it is package private, it can as well be called from another code path in the future. Hence, I believe the `synchronized` is required here.